packages/server/src/base/streamer.ts

```diff
@@ -0,0 +1,127 @@
+import { uuid } from '@botpress/messaging-base'
+import { Dispatcher, DispatchService, Logger } from '@botpress/messaging-engine'
+import axios, { AxiosRequestConfig } from 'axios'
+import clc from 'cli-color'
+import { backOff } from 'exponential-backoff'
+import yn from 'yn'
+import { ActionSource } from '../base/source'
+import { SocketEvents, SocketUserEvent } from '../socket/events'
+import { SocketService } from '../socket/service'
+import { WebhookService } from '../webhooks/service'
+
+const MAX_ATTEMPTS = 10
+
+export class Streamer {
+  private logger = new Logger('Stream')
+  private dispatcher!: StreamDispatcher
+  private destroyed: boolean = false
+
+  constructor(private dispatches: DispatchService, private sockets: SocketService, private webhooks: WebhookService) {}
+
+  async setup() {
+    this.sockets.events.on(SocketEvents.UserConnected, this.handleUserConnected.bind(this))
+    this.sockets.events.on(SocketEvents.UserDisconnected, this.handleUserDisconnected.bind(this))
+
+    this.dispatcher = await this.dispatches.create('dispatch_socket', StreamDispatcher)
+    this.dispatcher.on(StreamerDispatches.Message, this.handleDispatchMessage.bind(this))
+  }
+
+  async destroy() {
+    // TODO: we could close off all http connections here
+
+    this.destroyed = true
+  }
+
+  private async handleUserConnected({ userId }: SocketUserEvent) {
+    await this.dispatcher.subscribe(userId)
+  }
+
+  private async handleUserDisconnected({ userId }: SocketUserEvent) {
+    await this.dispatcher.unsubscribe(userId)
+  }
+
+  private async handleDispatchMessage(userId: uuid, { source, payload }: StreamerMessageDispatch) {
+    const sockets = this.sockets.listByUser(userId)
+
+    for (const socket of sockets) {
+      if (source !== socket.id) {
+        socket.send(payload)
+      }
+    }
+  }
+
+  async stream(type: string, data: any, clientId: uuid, userId?: uuid, source?: ActionSource) {
+    const payload = {
+      type,
+      data: {
+        clientId,
+        ...(userId ? { userId } : {}),
+        ...data
+      }
+    }
+
+    if (yn(process.env.LOGGING_ENABLED)) {
+      this.logger.debug(`${clc.blackBright(`[${clientId}]`)} ${clc.cyan(type)}`, payload)
+    }
+
+    if (userId) {
+      await this.dispatcher.publish(StreamerDispatches.Message, userId, { source: source?.socket?.id, payload })
```

Comment:
Maybe we should check if the user is connected through socket(s) before firing this event? This would prevent us from "disturbing" all the other servers when it might not be necessary. What do you think?

Reply:
I think you missed a part of the code. The handleUserConnected and handleUserDisconnected subscribe and unsubscribe the node to a particular user's events. So we in fact only disturb nodes where the user in connected